### ISSUE TITLE
Wrong Typescript version is used

### ISSUE THREAD

**Issue by exalted** [zed-industries/community#2278]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Consider the following project structure:

```
- package.json
- package-lock.json
- node_modules/typescript/lib/tsserver.js

- infrastructure/package.json
- infrastructure/package-lock.json
- infrastructure/node_modules/typescript/lib/tsserver.js
```

☝️ In essence, the above highlights the fact that this project has a "root" level and a subdirectory `infrastructure/` that contains yet another layer of node dependencies, which uses different versions of TypeScript:

Root-level:

```
project@1.0.0 $HOME/dev/project
└── typescript@3.7.4
```

Subdirectory:

```
project@0.1.0 $HOME/dev/project/infrastructure
└── typescript@5.2.2
```

Zed's issue with this setup is that when opened Zed from the root-level (i.e., `zed $HOME/dev/project`) here's the relevant parts of the `~/Library/Logs/Zed/Zed.log`:

```
2023-11-16T13:10:08 [INFO] build git repository ".git"
2023-11-16T13:10:08 [INFO] starting language server "typescript-language-server", path: "$HOME/dev/project", id: 5
2023-11-16T13:10:08 [INFO] starting language server "eslint", path: "$HOME/dev/project", id: 6
2023-11-16T13:10:08 [INFO] 6 unhandled notification window/logMessage:
{
  "type": 3,
  "message": "ESLint server running in node v18.15.0"
}
2023-11-16T13:10:08 [INFO] 5 unhandled notification window/logMessage:
{
  "type": 2,
  "message": "Using Typescript version (user-setting) 3.7.4 from path \"$HOME/dev/project/node_modules/typescript/lib/tsserver.js\""
}
2023-11-16T13:10:08 [INFO] 5 unhandled notification $/typescriptVersion:
{
  "version": "3.7.4",
  "source": "user-setting"
}
2023-11-16T13:10:08 [INFO] 6 unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file://$HOME/dev/project/infrastructure/src/ecs-app.ts"
  }
}
2023-11-16T13:10:08 [INFO] 6 unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file://$HOME/dev/project/infrastructure/src/index.ts"
  }
}
2023-11-16T13:10:08 [INFO] 6 unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file://$HOME/dev/project/infrastructure/src/mysql-stack.ts"
  }
}
2023-11-16T13:10:25 [ERROR] unexpected item event after pane was dropped
2023-11-16T13:10:25 [ERROR] unexpected item event after pane was dropped
2023-11-16T13:10:30 [INFO] 6 unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file://$HOME/dev/project/infrastructure/src/mysql-stack.ts"
  }
}
```

And this is when opening the subdirectory (i.e., `zed $HOME/dev/project/infrastructure`):

```
2023-11-16T13:11:26 [INFO] starting language server "typescript-language-server", path: "$HOME/dev/project/infrastructure", id: 7
2023-11-16T13:11:26 [INFO] starting language server "eslint", path: "$HOME/dev/project/infrastructure", id: 8
2023-11-16T13:11:26 [INFO] 8 unhandled notification window/logMessage:
{
  "type": 3,
  "message": "ESLint server running in node v18.15.0"
}
2023-11-16T13:11:26 [INFO] 7 unhandled notification window/logMessage:
{
  "type": 2,
  "message": "Using Typescript version (user-setting) 5.2.2 from path \"$HOME/dev/project/infrastructure/node_modules/typescript/lib/tsserver.js\""
}
2023-11-16T13:11:26 [INFO] 7 unhandled notification $/typescriptVersion:
{
  "version": "5.2.2",
  "source": "user-setting"
}
2023-11-16T13:11:26 [INFO] 8 unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file://$HOME/dev/project/infrastructure/src/mysql-stack.ts"
  }
}
2023-11-16T13:11:35 [INFO] 8 unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file://$HOME/dev/project/infrastructure/src/mysql-stack.ts"
  }
}
```

So, in short, for a file that is under the subdirectory (i.e., `$HOME/dev/project/infrastructure/src/mysql-stack.ts` in this example) root-level detects `Using Typescript version (user-setting) 3.7.4 from path \"$HOME/dev/project/node_modules/typescript/lib/tsserver.js\"` VS subdirectory detecting `Using Typescript version (user-setting) 5.2.2 from path \"$HOME/dev/project/infrastructure/node_modules/typescript/lib/tsserver.js\"`.

This is wrong, IMO, and Zed should use the "closest" TypeScript version to the file that is being edited and not where Zed's project root is.

This becomes an issue, because depending on the version of TypeScript detected, I am getting false-positive TypeScript errors, compared to when the correct version is used.

### Environment

Zed: v0.112.3 (stable)
OS: macOS 14.0.0
Memory: 64 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

**Comment by SomeoneToIgnore** [comment]
We have landed a few improvements to prettier in Zed Preview, would you mind testing it now?
Alternatively, I'd appreciate a working project we can test it on more.

**Comment by exalted** [comment]
@SomeoneToIgnore hmm… Thanks! Before I go ahead and do that though, nowhere in my issue report I was referring to Prettier. Are you sure we are on the same page? 🙏

**Comment by SomeoneToIgnore** [comment]
Apologies, I've meant "eslint" instead of prettier actually 🙂 
We did improve its detection capabilities so I would be interested to see if it fixed your case too.

**Comment by exalted** [comment]
@SomeoneToIgnore I retried using:

```
Zed: v0.116.1 (stable)
OS: macOS 14.1.1
Memory: 64 GiB
Architecture: aarch64
```

The issue is still there.

You asked, and I delivered a repo where you can reproduce this issue: https://github.com/exalted/bug_zed-typescript-detection/tree/main.

Let me know if I can help you further troubleshooting this.

**Comment by SomeoneToIgnore** [comment]
Thank you for the example: it looks like there's a mixture of two LSP servers in consideration: eslint and typescript-language-server.

The "Using Typescript version (user-setting) ..." line comes from the latter: https://github.com/typescript-language-server/typescript-language-server/blob/e9c0b117a9a5e273eb517dc0d337ecdf973f3dac/src/lsp-server.ts#L102 and is indeed something we did not work on yet properly: I've reported back on the eslint improvements.

So, we definitely have a configuration (?) bug for the typescript-language-server.

Most of the other log lines seem to be related to eslint. 
For it, Zed sends a `"workingDirectory": {"mode": "auto"}` and a workspace root abs path as a `"workspaceFolder"` config item into https://github.com/microsoft/vscode-eslint/blob/53712d973597e2d49be40a2821360065031594cf/%24shared/settings.ts#L173C1-L174

This will later be used by eslint to look up its configuration per file: https://github.com/microsoft/vscode-eslint/blob/53712d973597e2d49be40a2821360065031594cf/server/src/eslint.ts#L842
This part seems ok, since not really tied to TS version but rather its configs, which I expect eslint to resolve properly.

Except for that, I can see that there's neither eslint package nor its configuration is specified in the project: VSCode fails with 

```
[Info  - 21:34:50] ESLint library loaded from: /Users/someonetoignore/Downloads/bug_zed-typescript-detection/node_modules/eslint/lib/api.js
Uncaught exception received.
Error: No ESLint configuration found in /Users/someonetoignore/Downloads/bug_zed-typescript-detection/example-subdir.
    at CascadingConfigArrayFactory._finalizeConfigArray ....snip
```

errors and neither does Zed display any. 
That all looks expected, and I know Zed is capable of displaying eslint errors generally, e.g. in https://github.com/RomainLanz/adonis-bull-queue project.


-------------------

So, to sum it up, eslint language server works fine so far, but the way Zed configures/interacts with typescript-language-server is buggy, we should take care of that.

**Comment by exalted** [comment]
Hey, @SomeoneToIgnore . Thanks for the detailed explanation. I am still failing to understand why is this in any way related to eslint, but I guess you know Zed's internals better than I do, so it's all right. I trust your judgement there.

Looking forward to see fixes with the TS LSP. ✌️
